src/Illuminate/Testing/TestJson.php

```diff
@@ -0,0 +1,292 @@
+<?php
+
+namespace Illuminate\Testing;
+
+use Illuminate\Contracts\Support\Jsonable;
+use Illuminate\Support\Arr;
+use Illuminate\Support\Str;
+use Illuminate\Testing\Assert as PHPUnit;
+use JsonSerializable;
+
+class TestJson
+{
+    /**
+     * The original encoded json.
+     *
+     * @var \Illuminate\Contracts\Support\Jsonable|\JsonSerializable|array
+     */
+    public $json;
+
+    /**
+     * The decoded json contents.
+     *
+     * @var array|null
+     */
+    protected $decoded;
+
+    /**
+     * Create a new test json instance.
+     *
+     * @param  \Illuminate\Contracts\Support\Jsonable|array  $jsonable
+     * @return void
+     */
+    public function __construct($jsonable)
+    {
+        $this->json = $jsonable;
+
+        if ($jsonable instanceof JsonSerializable) {
+            $this->decoded = $jsonable->jsonSerialize();
+        } elseif ($jsonable instanceof Jsonable) {
+            $this->decoded = json_decode($jsonable->toJson(), true);
+        } elseif (is_array($jsonable)) {
+            $this->decoded = $jsonable;
+        } else {
+            $this->decoded = json_decode($jsonable, true);
+        }
+    }
+
+    /**
+     * Validate and return the decoded response JSON.
+     *
+     * @param  string|null  $key
+     * @return mixed
+     */
+    public function json($key = null)
+    {
+        return data_get($this->decoded, $key);
+    }
+
+    /**
+     * Assert that the response JSON has the expected count of items at the given key.
+     *
+     * @param  int  $count
+     * @param  string|null  $key
+     * @return $this
+     */
+    public function assertCount(int $count, $key = null)
```

Comment:
we probably want scalar types for both or neither params?

Reply:
Laravel 8 accepting scalar params other than `int` and `array`?

Reply:
Yeh, there's string used in some places.

Reply:
This is still outstanding.

Reply:
I honestly don't want to further complicate the PR for a style that:

- Is not used everywhere in Laravel
- Is not documented as the recommended structure, or advocated by @taylorotwell 
- Not being used in the class, for example `json($key = null)` should be `json(?string $key = null)` based on your request.